source/common/http/header_map_impl.cc

```diff
@@ -443,57 +434,79 @@ void HeaderMapImpl::addCopy(const LowerCaseString& key, uint64_t value) {
     return;
   }
   HeaderString new_key;
-  new_key.setCopy(key.get().c_str(), key.get().size());
+  new_key.setCopy(key.get());
   HeaderString new_value;
   new_value.setInteger(value);
   insertByKey(std::move(new_key), std::move(new_value));
   ASSERT(new_key.empty());   // NOLINT(bugprone-use-after-move)
   ASSERT(new_value.empty()); // NOLINT(bugprone-use-after-move)
 }
 
-void HeaderMapImpl::addCopy(const LowerCaseString& key, const std::string& value) {
+void HeaderMapImpl::addCopy(const LowerCaseString& key, absl::string_view value) {
   auto* entry = getExistingInline(key.get());
   if (entry != nullptr) {
     const uint64_t added_size = appendToHeader(entry->value(), value);
     addSize(added_size);
     return;
   }
   HeaderString new_key;
-  new_key.setCopy(key.get().c_str(), key.get().size());
+  new_key.setCopy(key.get());
   HeaderString new_value;
-  new_value.setCopy(value.c_str(), value.size());
+  new_value.setCopy(value);
   insertByKey(std::move(new_key), std::move(new_value));
   ASSERT(new_key.empty());   // NOLINT(bugprone-use-after-move)
   ASSERT(new_value.empty()); // NOLINT(bugprone-use-after-move)
 }
 
-void HeaderMapImpl::setReference(const LowerCaseString& key, const std::string& value) {
+void HeaderMapImpl::append(const LowerCaseString& key, absl::string_view value) {
+  bool found = false;
+  for (HeaderEntryImpl& header : headers_) {
+    if (header.key().getStringView() == key.get()) {
+      const uint64_t added_size = appendToHeader(header.value(), value);
+      addSize(added_size);
+      found = true;
```

Comment:
Do we need to `break;` here? or do we add the value to all occurrences of the key?

Reply:
Hm. I guess for `append` it makes sense to add the value to only a single occurrence of the header. It's unclear what I should do for the `setCopy` case as well... see discussion about the O(n) headers.

Reply:
Yeah this is a tough one. I could see going both ways with this. It's also related to the other point about whether we should be doing collapsing more aggressively or not that @asraa brought up. @alyssawilk @PiotrSikora any thoughts on this one?

Reply:
In the past I've used https://src.chromium.org/viewvc/chrome/trunk/src/net/http/http_util.cc#l368 (Chrome's HttpUtil::IsNonCoalescingHeader() function) as the source of truth for what to do here.

Some headers make sense to have multiple distinct values, so they show up as separate header lines, e.g. set-cookie, and cookie. I don't fully understand the others in the list from Chrome (Date, Expires, Last-Modified, Location, Retry-After, Www-Authenticate, Proxy-Authenticate, and Strict-Transport-Security). But I think the lesson is to have those as separate. 

It's also possible that this list is incomplete and there's a more appropriate canonical source in an http spec.

So I suggest append() for such headers in that list (including Cookie and Set-Cookie) should add a new HeaderEntry instance, and not mutate any existing ones.

I think for others, I think this argues they should be coalesced and comma-separated during ingestion, but I'm not advocating for that change in this PR. Instead I think asraa's suggestion of just updating the first one seems appropriate for this PR.

Reply:
I also want to point out here that ::get() returns only the first occurrence. AFAICT there is no way in the HeaderMap API to retrieve a list of values by name. You can only find extra values by iterating.

Reply:
Agreed on making as few functional changes in this PR as possible. Feel free to add TODOs, etc. for follow ups.

Reply:
Done. I made the functionality of `setCopy` and `appendCopy` the same as it used to be, and added a TODO to follow up on this. Both of these will only modify the first occurrence of the header, as the use cases did before.

Reply:
It's still missing a `break;`

Reply:
Yeah, or just return when you find it, and then you can drop the 'found' variable and call addCopy() unconditionally below if you get there.

Please also add
```
// TODO(#9221): converge on and document a policy for this.
```

Reply:
Hm... @marcomagdy .... I think you're viewing the outdated version. See all changes in the commits. The function calls a `getExisting` if the key exists, and appends to that key (it will only be the first occurrence). Otherwise `addCopy`.